git-protocol/src/fetch.rs

```diff
@@ -0,0 +1,493 @@
+
+        Action::Continue
+    }
+
+    fn negotiate(
+        &mut self,
+        refs: &[Ref],
+        args: &mut Arguments,
+        _: Option<&Response>,
+    ) -> io::Result<Action> {
+        for r in refs {
+            if let Some(WantHave { want, have }) = (self.filter_refs)(r)? {
+                args.want(want);
+                if let Some(oid) = have {
+                    args.have(oid);
+                }
+            }
+        }
+
+        for WantRef { name, have } in &self.opt.want_refs {
+            args.want_ref(name.as_bstr());
+            if let Some(oid) = have {
+                args.have(oid);
+            }
+        }
+
+        // send done, as we don't bother with further negotiation
+        Ok(Action::Cancel)
+    }
+}
+
+#[async_trait(?Send)]
+impl<F, P, O> Delegate for Fetch<F, P, O>
+where
+    F: Fn(&Ref) -> io::Result<Option<WantHave>>,
+    P: ReceivePack<Output = O>,
+{
+    async fn receive_pack(
+        &mut self,
+        pack: impl AsyncBufRead + Unpin + 'async_trait,
+        prog: impl Progress,
+        _: &[Ref],
+        resp: &Response,
+    ) -> io::Result<()> {
+        self.out.wanted_refs = resp
+            .wanted_refs()
+            .iter()
+            .map(|response::WantedRef { id, path }| Ref::Direct {
+                path: path.clone(),
+                object: *id,
+            })
+            .collect();
+        let out = self.receive_pack.receive_pack(pack, prog).await?;
+        self.out.pack = Some(out);
+
+        Ok(())
+    }
+}
+
+pub async fn fetch<F, P, O, R, W>(
+    opt: Options,
+    filter_refs: F,
+    receive_pack: P,
+    recv: R,
+    send: W,
+) -> io::Result<Outputs<O>>
+where
+    F: Fn(&Ref) -> io::Result<Option<WantHave>> + Send + 'static,
+    P: ReceivePack<Output = O> + Send + 'static,
+    O: Send + 'static,
+    R: AsyncRead + Unpin + Send + 'static,
+    W: AsyncWrite + Unpin + Send + 'static,
+{
+    use git_transport::{
+        client::git::{ConnectMode, Connection},
+        Protocol,
+    };
+
+    let conn = Connection::new(
+        recv,
+        send,
+        Protocol::V2,
+        opt.repo.clone(),
+        None::<(String, Option<u16>)>,
+        ConnectMode::Daemon,
+    );
+    blocking::unblock(move || {
+        let mut delegate = Fetch::new(opt, filter_refs, receive_pack);
+        future::block_on(git_protocol::fetch::fetch(
+            conn,
+            &mut delegate,
+            |_| unreachable!("credentials helper requested"),
+            progress::Discard,
+        ))
+        .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
+
+        Ok(delegate.out)
+    })
+    .await
+}
+
+/// Like [`futures_lite::io::Take`], but returns an error if and when the
+/// `limit` is exceeded.
+///
+/// Note that, unlike [`futures_lite::io::Take`], if a single poll reads past
+/// the limit, the excess bytes are _not_ discarded. Instead, an error is
+/// returned on the next poll.
+pub struct TryTake<R> {
```

Comment:
Actually… I was hoping to read that I can use plain `take()` to bail out if the packfile exceeds the given limit, because the bundle writer is streaming and recognises a premature EOF without further processing. Is that the case?

Reply:
Yes, it is streaming and will fail loudly if an entry can't be fully parsed. In the possible event of the amount of bytes taken coinciding with an entry boundary it will still fail as it wants to read a set amount of entries as advertised in the pack header, along with the trailer.
Thus partial packs will always result in an error.